src/node/mini_miner.cpp

```diff
@@ -0,0 +1,220 @@
+    for (const auto& txiter : cluster) {
+        if (to_be_replaced.find(txiter->GetTx().GetHash()) == to_be_replaced.end()) {
+            auto [mapiter, success] = entries_by_txid.emplace(std::make_pair(txiter->GetTx().GetHash(), MockMempoolEntry(txiter)));
+            assert(success);
+            entries.push_back(mapiter);
+        } else {
+            auto outpoints_it = outpoints_needed_by_txid.find(txiter->GetTx().GetHash());
+            if (outpoints_it != outpoints_needed_by_txid.end()) {
+                for (const auto& outpoint : outpoints_it->second) {
+                    this->bump_fees.emplace(std::make_pair(outpoint, 0));
+                }
+            }
+        }
+    }
+
+    // Remove the to-be-replaced transactions and build the descendant_set_by_txid cache.
+    for (const auto& txiter : cluster) {
+        const auto& txid = txiter->GetTx().GetHash();
+        // Cache descendants for future use. Unlike the real mempool, a descendant MockMempoolEntry
+        // will not exist without its ancestor MockMempoolEntry, so these sets won't be invalidated.
+        std::vector<MockEntryMap::iterator> cached_descendants;
+        cached_descendants.emplace_back(entries_by_txid.find(txid));
+        // If a tx is to-be-replaced, remove any of its descendants so they can't fee-bump anything.
+        // this case should be rare as the wallet won't normally attempt to replace transactions
+        // with descendants.
+        const bool remove = to_be_replaced.find(txid) != to_be_replaced.end();
+        CTxMemPool::setEntries descendants;
+        mempool.CalculateDescendants(txiter, descendants);
+
+        for (const auto& desc_txiter : descendants) {
+            auto desc_it{entries_by_txid.find(desc_txiter->GetTx().GetHash())};
+            if (desc_it != entries_by_txid.end()) {
+                if (remove) {
+                    entries_by_txid.erase(desc_it);
+                } else {
+                    cached_descendants.push_back(desc_it);
+                }
+            }
+        }
+        if (!remove) descendant_set_by_txid.emplace(std::make_pair(txid, cached_descendants));
+    }
+    // Release the mempool lock; we now have all the information we need for a subset of the entries
+    // we care about. We will solely operate on the MockMempoolEntry map from now on.
+    assert(entries.size() == entries_by_txid.size());
+    assert(entries.size() == descendant_set_by_txid.size());
+    assert(in_block.empty());
+}
+
+// Compare by ancestor feerate, then iterator
+struct AncestorFeerateComparator
+{
+    template<typename I>
+    bool operator()(const I& a, const I& b) const {
+        const CFeeRate a_feerate(a->second.GetModFeesWithAncestors(), a->second.GetSizeWithAncestors());
+        const CFeeRate b_feerate(b->second.GetModFeesWithAncestors(), b->second.GetSizeWithAncestors());
+        if (a_feerate != b_feerate) {
+            return a_feerate > b_feerate;
+        }
+        return &(*a) > &(*b);
```

Comment:
what is the significance here? why not just have `return a_feerate >= b_feerate` or something like that?

Reply:
Added a comment